plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,648 @@
+				    FuCapCmd *req,
+				    gulong delay_us,
+				    GError **error)
+{
+	FuCapCmdHidReport report;
+	guint elapsed_ms = 0;
+
+	/*First two bytes are report id.*/
+	report.report_id = GINT16_TO_LE(FU_SYNAPTICS_CAPE_DEVICE_GOLEM_REPORT_ID);
+
+	memcpy(&report.cmd, req, sizeof(report.cmd));
+
+	/*Set data length to MAX for any GET commands*/
+	if (FU_SYNAPTICS_CMD_GET_FLAG & report.cmd.cmd_id) {
+		report.cmd.data_len = GUINT32_TO_LE(FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN);
+	}
+	report.cmd.data_len = GUINT32_TO_LE(report.cmd.data_len);
+	report.cmd.cmd_id = GUINT32_TO_LE(report.cmd.cmd_id);
+	report.cmd.module_id = GUINT32_TO_LE(report.cmd.module_id);
+
+	if (!fu_synaptics_cape_device_set_report(self,
+						 (const guint8 *)&report,
+						 sizeof(report),
+						 error)) {
+		g_prefix_error(error, "failed to send: ");
+		return FALSE;
+	}
+	if (delay_us > 0)
+		g_usleep(delay_us);
+
+	/* wait for the command to complete */
+	for (; elapsed_ms < FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_TIMEOUT;
+	     elapsed_ms += FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL) {
+		if (!fu_synaptics_cape_device_get_report(self,
+							 (guint8 *)&report,
+							 sizeof(report),
+							 error))
+			return FALSE;
+		if (report.cmd.reply)
+			break;
+		g_usleep(FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL * 1000);
+	}
+
+	report.cmd.data_len = GUINT32_FROM_LE(report.cmd.data_len);
+
+	/* copy returned data if it is GET command*/
+	if (FU_SYNAPTICS_CMD_GET_FLAG & report.cmd.cmd_id && report.cmd.reply) {
+		memcpy(req, &report.cmd, sizeof(report.cmd));
+	}
+
+	return fu_synaptics_cape_device_rc_set_error(&report.cmd, error);
+}
+
+gboolean
+fu_synaptics_cape_device_setup_version(FuSynapticsCapeDevice *self, GError **error)
+{
+	guint32 version_raw;
+	FuCapCmd cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	cmd.cmd_id = GUINT32_TO_LE(FU_SYNAPTICS_CMD_GET_VERSION);
+	cmd.module_id = FU_SYNAPTICS_CAPE_CMD_APP_ID('C', 'T', 'R', 'L');
+	cmd.data_len = 4;
+
+	fu_synaptics_cape_device_sendcmd_ex(self, &cmd, 0, error);
+	if (self->version) {
+		g_free(self->version);
+	}
+	version_raw = (cmd.data[0] & 0xff) << 24 | (cmd.data[1] & 0xff) << 16 |
+		      (cmd.data[2] & 0x8) << 24 | (cmd.data[3] & 0xff);
+
+	self->version = fu_common_version_from_uint32(version_raw, FWUPD_VERSION_FORMAT_QUAD);
+	fu_device_set_version(FU_DEVICE(self), self->version);
+	fu_device_set_version_raw(FU_DEVICE(self), version_raw);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+
+	return TRUE;
+}
+
+static void
+fu_synaptics_cape_device_init(FuSynapticsCapeDevice *self)
+{
+	fu_device_add_icon(FU_DEVICE(self), "audio-card");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_NEEDS_REBOOT);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_QUAD);
+	fu_device_set_install_duration(FU_DEVICE(self), 3); /* seconds */
+	fu_device_add_protocol(FU_DEVICE(self), "com.synaptics.cape");
```

Comment:
do you want this added on the LVFS? If so I just need to know:

 * Can the device do attestation post-update, i.e. read out a firmware checksum?
 * Is the firmware payload **always** signed?
 * What kind of devices will this be used for? e.g. mouse/keyboard/camera etc?

Reply:
Thank you for your code review. Yes, I want to this added on the LVFS. 

- The device will read out a firmware CRC32 number.
- The payload is not signed. 
- This is an USB audio codec, but the same protocol can be used on CAPE based Audio DSP chips.

Reply:
> Yes, I want to this added on the LVFS.

I've done this for you.

> The device will read out a firmware CRC32 number.

Can you add that as part of the ->setup() process? then add it as a checksum using fu_device_add_checksum?

Reply:
It looks like doable, but not now today. We can add this feature in future.